tempesta_fw/sched/tfw_sched_ratio.c

```diff
@@ -0,0 +1,467 @@
+ * Return 1 to indicate that all ratios/weights are equal.
+ * Return 0 to indicate that ratios are calculated by individial weights.
+ */
+static int
+tfw_sched_ratio_static(TfwSrvGroup *sg, int *ratio)
+{
+	size_t s;
+	TfwRatioSrvList *sl = sg->sched_data;
+	int diff = 0, wequal, weight[sl->srv_n];
+
+	wequal = sl->srvs[0].srv->weight;
+	for (s = 0; s < sl->srv_n; ++s) {
+		weight[s] = sl->srvs[s].srv->weight;
+		diff |= (wequal != weight[s]);
+	}
+	if (!diff) {
+		tfw_sched_ratio_equal(sg, ratio);
+		return 1;
+	}
+
+	tfw_sched_ratio_calc(weight, ratio, sl->srv_n);
+
+	return 0;
+}
+
+/**
+ * Add a connection and a server, if new, to the scheduler.
+ * Called at configuration stage, no synchronization is required.
+ */
+static void
+tfw_sched_ratio_add_conn(TfwSrvGroup *sg, TfwServer *srv, TfwSrvConn *srv_conn)
+{
+	size_t s, c;
+	TfwRatioSrv *cl;
+	TfwRatioSrvList *sl = sg->sched_data;
+
+	BUG_ON(!sl);
+
+	for (s = 0; s < sl->srv_n; ++s)
+		if (sl->srvs[s].srv == srv)
+			break;
+	BUG_ON(s == TFW_SG_MAX_SRV);
+
+	cl = &sl->srvs[s];
+
+	if (s == sl->srv_n) {
+		int nodiff, ratio[sl->srv_n + 1];
+		TfwRatioRated *rated = sl->rated;
+
+		cl->srv = srv;
+		++sl->srv_n;
+
+		if (sg->flags & TFW_SG_F_SCHED_RATIO_STATIC)
+			nodiff = tfw_sched_ratio_static(sg, ratio);
+		else if (sg->flags & TFW_SG_F_SCHED_RATIO_DYNAMIC)
+			nodiff = tfw_sched_ratio_equal(sg, ratio);
+		else
+			BUG();
+
+		if (nodiff)
+			tfw_sched_fill_equal(sg, ratio, rated);
+		else
+			tfw_sched_fill_rated(sg, ratio, rated);
+	}
+
+	for (c = 0; c < cl->conn_n; ++c) {
+		if (cl->conns[c] == srv_conn) {
+			TFW_WARN("sched=[%s]: attempt to add an existing "
+				 "connection: sg=[%s] srv=[%zd] conn=[%zd]\n",
+				 sg->sched->name, sg->name, s, c);
+			return;
+		}
+	}
+	BUG_ON(c == TFW_SRV_MAX_CONN);
+
+	cl->conns[c] = srv_conn;
+	++cl->conn_n;
+}
+
+/**
+ * On each subsequent call the function returns the next available
+ * connection to one of the servers in the group. Connections to a
+ * server are rotated in pure round-robin manner. The server is chosen
+ * according to its current weight. Servers with more weight are chosen
+ * more often than servers with less weight. 
+ * Dead connections and servers w/o live connections are skipped.
+ */
+static TfwSrvConn *
+tfw_sched_ratio_get_srv_conn(TfwMsg *msg, TfwSrvGroup *sg)
+{
+	size_t s, c, n;
+	uint64_t idx;
+	TfwSrvConn *srv_conn;
+	TfwRatioSrvList *sl = sg->sched_data;
+	TfwRatioSrv *cl, *cl_down = NULL;
+	TfwRatioRated *rated;
+
+	BUG_ON(!sl);
+
+	rcu_read_lock();
+	rated = rcu_dereference(sl->rated);
+
+	for (s = 0, n = 0; s < sl->srv_n; ++s) {
+		do {
+			idx = atomic64_inc_return(&sl->counter);
+			cl = rated->list[idx % rated->listsz];
+			if (likely(cl != cl_down))
+				break;
+		} while (++n < rated->listsz);
+
+		if (unlikely(n == rated->listsz))
+			continue;
+
+		for (c = 0; c < cl->conn_n; ++c) {
+			idx = atomic64_inc_return(&cl->counter);
+			srv_conn = cl->conns[idx % cl->conn_n];
```

Comment:
With #76 in mind `conns` can be dynamically allocated rather declared with static size as currently. It seems original RR algorithm is the most straightforward to satisfy #76 requirements, since it basically doesn't scan all the servers/connection as for example hash scheduler.

BTW the function is pretty complex, so a good comments are required describing the complex loops, e.g. the do-while.